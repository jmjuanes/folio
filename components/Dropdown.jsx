import React from "react";
import classNames from "classnames";

import {CheckIcon, ExternalLinkIcon} from "./Icons.jsx";

export const DropdownSeparator = () => (
    <div className="bg-neutral-200 w-full h-px my-2" />
);

export const DropdownGroup = props => (
    <div className="text-xs mb-1 px-2 text-neutral-600 select-none">
        {props.title}
    </div>
);

export const DropdownItem = props => {
    const classList = classNames({
        "flex items-center gap-2 rounded-md px-3 py-2 select-none": true,
        "bg-neutral-950 text-white": props.active && !props.disabled,
        "hover:bg-neutral-100 cursor-pointer": !props.active && !props.disabled,
        "o-70 cursor-not-allowed": props.disabled,
    });
    const handleClick = () => {
        if (!props.disabled && typeof props.onClick === "function") {
            if (document.activeElement && document.activeElement !== document.body) {
                document.activeElement.blur();
            }
            return props.onClick();
        }
    };
    return (
        <div className={classList} tabIndex="0" onClick={handleClick}>
            <div className="flex text-lg items-center">
                {props.icon}
            </div>
            <div className="flex items-center text-sm">
                <span>{props.text}</span>
            </div>
        </div>
    );
};

DropdownItem.defaultProps = {
    icon: null,
    text: null,
    active: false,
    disabled: false,
    onClick: null,
};

export const DropdownLinkItem = props => {
    const classList = classNames({
        "flex items-center gap-2 rounded-md px-3 py-2 select-none": true,
        "hover:bg-neutral-100 cursor-pointer no-underline": true,
    });
    return (
        <a className={classList} href={props.url} target="_blank">
            <div className="flex text-lg items-center">
                <ExternalLinkIcon />
            </div>
            <div className="flex items-center text-sm">
                <span>{props.text}</span>
            </div>
        </a>
    );
};

DropdownLinkItem.defaultProps = {
    url: "",
    text: "",
};

export const DropdownCheckItem = props => (
    <div className="flex items-center gap-2 rounded-md px-3 py-2 select-none hover:bg-neutral-100 cursor-pointer" onClick={props.onClick}>
        <div className="flex text-lg items-center">
            {props.icon}
        </div>
        <div className="flex items-center text-sm">
            <span>{props.text}</span>
        </div>
        {props.active && (
            <div className="ml-auto flex items-center text-sm">
                <CheckIcon />
            </div>
        )}
    </div>
);

DropdownCheckItem.defaultProps = {
    active: false,
    icon: null,
    text: null,
    onClick: null,
};

export const Dropdown = props => (
    <div className={classNames(props.className, "absolute mt-1 z-5")}>
        <div className="bg-white shadow-md w-56 p-1 rounded-lg flex flex-col gap-0 border border-neutral-200">
            {props.children}
        </div>
    </div>
);

Dropdown.defaultProps = {
    className: "top-0 left-0",
};
