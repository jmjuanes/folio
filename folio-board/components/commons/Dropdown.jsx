import React from "react";
import classNames from "classnames";

export const DropdownItem = props => {
    const classList = classNames({
        "d:flex items:center gap:2 r:md px:3 py:2 select:none": true,
        "bg:light-200:hover cursor:pointer": !props.disabled,
        "o:80 cursor:not-allowed": props.disabled,
    });
    const handleClick = () => {
        if (!props.disabled && typeof props.onClick === "function") {
            if (document.activeElement && document.activeElement !== document.body) {
                document.activeElement.blur();
            }
            return props.onClick();
        }
    };
    return (
        <div className={classList} tabIndex="0" onClick={handleClick}>
            <div className="d:flex text:lg items:center text:dark-700">
                {props.icon}
            </div>
            <div className="d:flex items:center text:sm text:dark-700">
                <span>{props.text}</span>
            </div>
        </div>
    );
};

DropdownItem.defaultProps = {
    icon: null,
    text: null,
    disabled: false,
    onClick: null,
};

export const Dropdown = props => (
    <div className={classNames(props.className, "position:absolute mt:1")}>
        <div className="bg:white shadow:md w:56 p:3 r:lg d:flex flex:col gap:0">
            {props.children}
        </div>
    </div>
);

Dropdown.defaultProps = {
    className: "top:0 left:0",
};
