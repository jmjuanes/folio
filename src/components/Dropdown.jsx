import React from "react";
import classNames from "classnames";

import {CheckIcon, ExternalLinkIcon} from "./Icons.jsx";

export const DropdownSeparator = () => (
    <div className="bg-gray-200 h-px w-full my-2" />
);

export const DropdownGroup = props => (
    <div className="text-xs mb-2 text-gray-500 select-none">
        {props.title}
    </div>
);

export const DropdownItem = props => {
    const classList = classNames({
        "flex items-center gap-2 rounded-md px-3 py-2 select-none": true,
        "hover:bg-gray-200 cursor-pointer": !props.disabled,
        "o-80 cursor-not-allowed": props.disabled,
    });
    const handleClick = () => {
        if (!props.disabled && typeof props.onClick === "function") {
            if (document.activeElement && document.activeElement !== document.body) {
                document.activeElement.blur();
            }
            return props.onClick();
        }
    };
    return (
        <div className={classList} tabIndex="0" onClick={handleClick}>
            <div className="flex text-lg items-center text-gray-700">
                {props.icon}
            </div>
            <div className="flex items-center text-sm text-gray-700">
                <span>{props.text}</span>
            </div>
        </div>
    );
};

DropdownItem.defaultProps = {
    icon: null,
    text: null,
    disabled: false,
    onClick: null,
};

export const DropdownLinkItem = props => {
    const classList = classNames({
        "flex items-center gap-2 rounded-md px-3 py-2 select-none": true,
        "hover:bg-gray-200 cursor-pointer no-underline": true,
    });
    return (
        <a className={classList} href={props.url} target="_blank">
            <div className="flex text-lg items-center text-gray-700">
                <ExternalLinkIcon />
            </div>
            <div className="flex items-center text-sm text-gray-700">
                <span>{props.text}</span>
            </div>
        </a>
    );
};

DropdownLinkItem.defaultProps = {
    url: "",
    text: "",
};

export const DropdownCheckItem = props => (
    <div className="flex items-center gap-2 rounded-md px-3 py-2 select-none hover:bg-gray-200 cursor-pointer" onClick={props.onClick}>
        <div className="flex text-lg items-center text-gray-700">
            {props.icon}
        </div>
        <div className="flex items-center text-sm text-gray-700">
            <span>{props.text}</span>
        </div>
        {props.active && (
            <div className="ml-auto flex items-center text-sm text-gray-800">
                <CheckIcon />
            </div>
        )}
    </div>
);

DropdownCheckItem.defaultProps = {
    active: false,
    icon: null,
    text: null,
    onClick: null,
};

export const Dropdown = props => (
    <div className={classNames(props.className, "absolute mt-1")}>
        <div className="bg-white shadow-md w-56 p-3 rounded-lg flex flex-col gap-0">
            {props.children}
        </div>
    </div>
);

Dropdown.defaultProps = {
    className: "top-0 left-0",
};
