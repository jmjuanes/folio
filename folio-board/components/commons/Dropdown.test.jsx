import TestRenderer from "react-test-renderer";
import {Dropdown, DropdownItem} from "./Dropdown.jsx";

describe("Dropdown", () => {
    it("should render", () => {
        const testRenderer = TestRenderer.create((
            <Dropdown
                items={[
                    {text: "Item 1"},
                    {text: "Item 2"},
                ]}
            />
        ));

        expect(testRenderer.toJSON()).toMatchSnapshot();
    });

    it("should render the provided items", () => {
        const items = [
            {text: "Item 1"},
            {text: "Item 2"},
        ];
        const testRenderer = TestRenderer.create(<Dropdown items={items} />);
        const testInstance = testRenderer.root;
        const childItems = testInstance.findAllByType(DropdownItem);

        expect(childItems[0].props.text).toEqual("Item 1");
        expect(childItems[1].props.text).toEqual("Item 2");
    });

    it("should call the provided onClick listener", () => {
        const mockFn = jest.fn();
        const items = [
            {text: "Item 1", onClick: mockFn},
        ];
        const testRenderer = TestRenderer.create(<Dropdown items={items} />);

        // Run the onClick method
        testRenderer.root.findByType(DropdownItem).props.onClick();

        expect(mockFn).toHaveBeenCalled();
    });
});
