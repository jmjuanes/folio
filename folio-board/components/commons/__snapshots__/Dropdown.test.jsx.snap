// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown should render 1`] = `
<div
  className="top:0 left:0 position:absolute mt:1"
>
  <div
    className="bg:white shadow:md w:56 p:3 r:lg d:flex flex:col gap:0"
  >
    <div
      className="d:flex items:center gap:2 r:md px:3 py:2 select:none bg:light-200:hover cursor:pointer"
      onClick={[Function]}
      tabIndex="0"
    >
      <div
        className="d:flex text:lg items:center text:dark-700"
      />
      <div
        className="d:flex items:center text:sm text:dark-700"
      >
        <span>
          Item 1
        </span>
      </div>
    </div>
    <div
      className="d:flex items:center gap:2 r:md px:3 py:2 select:none bg:light-200:hover cursor:pointer"
      onClick={[Function]}
      tabIndex="0"
    >
      <div
        className="d:flex text:lg items:center text:dark-700"
      />
      <div
        className="d:flex items:center text:sm text:dark-700"
      >
        <span>
          Item 2
        </span>
      </div>
    </div>
  </div>
</div>
`;
